bot/seasons/evergreen/minesweeper.py

```diff
@@ -0,0 +1,212 @@
+import typing
+from random import random
+
+from discord.ext import commands
+
+GameBoard = typing.List[typing.List[typing.Union[str, int]]]
+DictOfGames = typing.Dict[int, typing.Dict]
+
+
+class CordConverter(commands.Converter):
+    """Converter for cords."""
+
+    async def convert(self, ctx, cord: str) -> typing.Tuple[int, int]:
+        """Take in a cord string and turn it into x, y"""
+        if not 2 <= len(cord) <= 3:
+            raise commands.ArgumentParsingError()
+        value1 = cord[0]
+        value2 = cord[1:]
+        if not value2.isdigit():
+            raise commands.ArgumentParsingError()
+        x = ord(value1) - 97
+        y = int(value2) - 1
+        if (not 0 <= x <= 9) or (not 0 <= y <= 9):
+            raise commands.ArgumentParsingError()
+        return x, y
+
+
+class Minesweeper(commands.Cog):
+    """Play a game of minesweeper."""
+
+    def __init__(self, bot: commands.Bot) -> None:
+        self.games: DictOfGames = {}  # Store the currently running games
+
+    @staticmethod
+    def is_bomb(cell: typing.Union[str, int]) -> int:
+        """Returns 1 if `cell` is a bomb if not 0"""
+        return cell == "bomb"
+
+    def generate_board(self, bomb_chance: float) -> GameBoard:
+        """Generate a 2d array for the board."""
+        board: GameBoard = [["bomb" if random() <= bomb_chance else "number" for _ in range(10)] for _ in range(10)]
+        for y, row in enumerate(board):
+            for x, cell in enumerate(row):
+                if cell == "number":
+                    # calculate bombs near it
+                    bombs = 0
+                    for x_ in [x - 1, x, x + 1]:
+                        for y_ in [y - 1, y, y + 1]:
+                            if x_ != -1 and x_ != 10 and y_ != -1 and y_ != 10 and board[y_][x_] == "bomb":
+                                bombs += 1
+
+                    board[y][x] = bombs
+        return board
+
+    @staticmethod
+    def format_for_discord(board: GameBoard) -> str:
+        """Format the board to a string for discord."""
+        mapping = {
+            0: ":stop_button:",
+            1: ":one:",
+            2: ":two:",
+            3: ":three:",
+            4: ":four:",
+            5: ":five:",
+            6: ":six:",
+            7: ":seven:",
+            8: ":eight:",
+            9: ":nine:",
+            10: ":keycap_ten:",
+            "bomb": ":bomb:",
+            "hidden": ":grey_question:",
+            "flag": ":triangular_flag_on_post:"
```

Comment:
only the last 3? would it not be better to just keep it as it is then?

Reply:
Hmm yeah, using an enum may make it more convoluted... You could just keep them in this dict but change the keys from string to enum instead of relying on enum values directly for the emoji conversions.

My thought was that using strings instead of enum constants is more prone to programmer errors. Enums work well with static type checkers so it'll be easier to spot when an incorrect value is used.

Reply:
I am not sure I completely understand what you mean? have an Enum class for the emojis and then use that as the values?

Reply:
```py
class Cell(Enum):
    BOMB
    HIDDEN
    FLAG

{
    10: ":keycap_ten:",
    Cell.BOMB = ":bomb:"
}

board[y][x] = Cell.BOMB
```

Still not sure if it's a good idea.

Reply:
hm, idk. I think it good with just strings. are there any performance improvements using that?

Reply:
Doubtful. I wasn't concerned with performance to begin with.

Reply:
then, I think it's okay the way it is unless somebody wants to add that.